### PR TITLE
Biased Random Walks and Node2Vec implementation

### PR DESCRIPTION
**Pull request by ChuckHastings**
Replaces #4499 
Closes #4499 

@G-Cornett implemented biased random walks and node2vec during his internship.  This PR includes his changes to implement those algorithms.

### CODE REVIEW

File: cpp/src/prims/detail/sample_and_compute_local_nbr_indices.cuh [hunk 1]
```diff
@@ -392,11 +392,11 @@ compute_unique_keys(raft::handle_t const& handle,
       cuda::proclaim_return_type<size_t>(
         [unique_key_first = get_dataframe_buffer_begin(aggregate_local_frontier_unique_keys) +
                             local_frontier_unique_key_displacements[i],
-         num_unique_keys = local_frontier_unique_key_sizes[i]] __device__(key_t key) {
+         unique_key_last = get_dataframe_buffer_begin(aggregate_local_frontier_unique_keys) +
+                           local_frontier_unique_key_displacements[i] +
+                           local_frontier_unique_key_sizes[i]] __device__(key_t key) {
```
Comment (ChuckHastings) [inline cpp/src/prims/detail/sample_and_compute_local_nbr_indices.cuh:397]: Note, passing num_unique_keys was failing on CUDA 11.8.  Not sure why, seems like perhaps a compiler bug.  Passing first and last worked on both CUDA 11.8 and CUDA 12.

Comment (jnke2016) [reply cpp/src/prims/detail/sample_and_compute_local_nbr_indices.cuh:397]: I have seen the issue especially in lambda functions hence now I often pass first and last

**Review by jnke2016 — APPROVED** [review]

**Review by naimnv — APPROVED** [review]

**Review by seunghwak — APPROVED** [review]
LGTM

**Pull request merged**